src/OpticalFlowEstimator.cpp

```diff
@@ -0,0 +1,598 @@
+{
+    return std::hypot(img_size.width/2.0, img_size.height/2.0)
+         / std::tan(fov / 2.0);
+}
+
+void OpticalFlowEstimator::getYPR(const tf2::Quaternion& orientation,
+                                  double& y,
+                                  double& p,
+                                  double& r)
+{
+    tf2::Matrix3x3 matrix;
+    matrix.setRotation(orientation);
+    matrix.getEulerYPR(y, p, r);
+}
+
+void OpticalFlowEstimator::processImage(const cv::gpu::GpuMat& image,
+                                        const cv::gpu::GpuMat& gray_image,
+                                        const tf2::Quaternion& orientation,
+                                        const ros::Time& time,
+                                        double height,
+                                        bool debug) const
+{
+    // Find vectors from image
+    std::vector<cv::Point2f> tails;
+    std::vector<cv::Point2f> heads;
+    std::vector<uchar> status;
+    findFeatureVectors(image,
+                       gray_image,
+                       last_scaled_image_,
+                       last_scaled_grayscale_image_,
+                       tails,
+                       heads,
+                       status,
+                       debug);
+
+    // Calculate the average vector
+    cv::Point2f average_vec = findAverageVector(
+            tails,
+            heads,
+            status,
+            flow_estimator_settings_.x_cutoff_region_velocity_measurement,
+            flow_estimator_settings_.y_cutoff_region_velocity_measurement,
+            target_size_);
+
+    // Publish debugging image with average vector drawn
+    if (debug && debug_settings_.debug_average_vector_image) {
+        cv::Mat arrow_image;
+        last_scaled_image_.download(arrow_image);
+
+        const cv::Point2f start_point(target_size_.width / 2.0,
+                                      target_size_.height / 2.0);
+
+        const cv::Point2f end_point(average_vec.x + start_point.x,
+                                    average_vec.y + start_point.y);
+
+        drawArrows(arrow_image,
+                   { start_point },
+                   { end_point },
+                   { 1 },
+                   cv::Scalar(255, 0, 0));
+
+        const cv_bridge::CvImage cv_image {
+            std_msgs::Header(),
+            sensor_msgs::image_encodings::RGBA8,
+            arrow_image
+        };
+
+        debug_average_velocity_vector_image_pub_.publish(
+                cv_image.toImageMsg());
+    }
+
+    const geometry_msgs::TwistWithCovarianceStamped velocity = estimateVelocity(
+            average_vec,
+            orientation,
+            last_orientation_,
+            height,
+            time);
+
+    if (!std::isfinite(velocity.twist.twist.linear.x)
+     || !std::isfinite(velocity.twist.twist.linear.y)
+     || !std::isfinite(velocity.twist.covariance[0])    // variance of x
+     || !std::isfinite(velocity.twist.covariance[1])    // covariance of x & y
```

Comment:
Why are we checking the covariance of x and y if we never set it?

Reply:
Because those are the values that are relevant to the output, and we might end up setting them in the future